soci/store/store.go

```diff
@@ -0,0 +1,337 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"path/filepath"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/errdef"
+)
+
+const (
+	DefaultContainerdSocket = "/run/containerd/containerd.sock"
+)
+
+// BasicStore describes the functionality common to oras-go oci.Store, oras-go memory.Store, and containerd ContentStore
+type BasicStore interface {
+	Exists(ctx context.Context, target ocispec.Descriptor) (bool, error)
+	Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error)
+	Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error
+}
+
+// Store extends BasicStore with functionality that in not present in some BasicStore
+// implementations and may be stubbed in some Store implementations
+type Store interface {
+	BasicStore
+	Label(ctx context.Context, target ocispec.Descriptor, label string, value string) error
+	Delete(ctx context.Context, dgst digest.Digest) error
+}
+
+type ContentStoreType string
+
+const (
+	ContainerdContentStoreType ContentStoreType = "containerd"
+	SociContentStoreType       ContentStoreType = "soci"
+	DefaultContentStoreType    ContentStoreType = SociContentStoreType
+)
+
+// GetContentStoreTypes returns a slice of all supported content store types
+func GetContentStoreTypes() []ContentStoreType {
```

Comment:
nit to be more go-idomatic:

>Go doesn't provide automatic support for getters and setters. There's nothing wrong with providing getters and setters yourself, and it's often appropriate to do so, but it's neither idiomatic nor necessary to put Get into the getter's name.

https://go.dev/doc/effective_go

```suggestion
func ContentStoreTypes() []ContentStoreType {
```